### ISSUE TITLE
Build failure: FreeCAD

### ISSUE THREAD

**Issue by Lordraven19** [NixOS/nixpkgs#263452]
### Steps To Reproduce

1. build FreeCAD on NixOS Unstable (or likely 23.11, when released)

### Build log

```
       … while evaluating derivation 'freecad-0.21.1'
         whose name attribute is located at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/stdenv/generic/make-derivation.nix:300:7

       … while evaluating attribute 'buildInputs' of derivation 'freecad-0.21.1'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/stdenv/generic/make-derivation.nix:347:7:

          346|       depsHostHost                = lib.elemAt (lib.elemAt dependencies 1) 0;
          347|       buildInputs                 = lib.elemAt (lib.elemAt dependencies 1) 1;
             |       ^
          348|       depsTargetTarget            = lib.elemAt (lib.elemAt dependencies 2) 0;

       … while calling 'getOutput'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/attrsets.nix:962:23:

          961|   */
          962|   getOutput = output: pkg:
             |                       ^
          963|     if ! pkg ? outputSpecified || ! pkg.outputSpecified

       … while calling anonymous lambda

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/stdenv/generic/make-derivation.nix:246:13:

          245|       (map (drv: drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHost" depsHostHost))
          246|       (map (drv: drv.__spliced.hostTarget or drv) (checkDependencyList "buildInputs" buildInputs'))
             |             ^
          247|     ]

       … while calling anonymous lambda

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/lists.nix:165:29:

          164|   */
          165|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          166|

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/lists.nix:165:32:

          164|   */
          165|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          166|

       … while calling anonymous lambda

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/stdenv/generic/make-derivation.nix:213:81:

          212|   checkDependencyList = checkDependencyList' [];
          213|   checkDependencyList' = positions: name: deps: lib.flip lib.imap1 deps (index: dep:
             |                                                                                 ^
          214|     if lib.isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/stdenv/generic/make-derivation.nix:214:8:

          213|   checkDependencyList' = positions: name: deps: lib.flip lib.imap1 deps (index: dep:
          214|     if lib.isDerivation dep || dep == null || builtins.isString dep || builtins.isPath dep then dep
             |        ^
          215|     else if lib.isList dep then checkDependencyList' ([index] ++ positions) name dep

       … while calling 'isDerivation'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/attrsets.nix:670:5:

          669|     # Value to check.
          670|     value: value.type or null == "derivation";
             |     ^
          671|

       … while calling 'func'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/interpreters/python/passthrufun.nix:26:20:

           25|       valid = value: pythonPackages.hasPythonModule value || providesSetupHook value || lib.elem value exceptions;
           26|       func = name: value:
             |                    ^
           27|         if lib.isDerivation value then

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/interpreters/python/passthrufun.nix:27:12:

           26|       func = name: value:
           27|         if lib.isDerivation value then
             |            ^
           28|           lib.extendDerivation (valid value || throw "${name} should use `buildPythonPackage` or `toPythonModule` if it is to be part of the Python packages set.") {} value

       … while calling 'isDerivation'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/attrsets.nix:670:5:

          669|     # Value to check.
          670|     value: value.type or null == "derivation";
             |     ^
          671|

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/top-level/python-packages.nix:10877:13:

        10876|
        10877|   pyside2 = toPythonModule (callPackage ../development/python-modules/pyside2 {
             |             ^
        10878|     inherit (pkgs) cmake ninja qt5;

       … while calling 'toPythonModule'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/interpreters/python/python-packages-base.nix:68:20:

           67|   # Convert derivation to a Python module.
           68|   toPythonModule = drv:
             |                    ^
           69|     drv.overrideAttrs( oldAttrs: {

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/top-level/python-packages.nix:10877:29:

        10876|
        10877|   pyside2 = toPythonModule (callPackage ../development/python-modules/pyside2 {
             |                             ^
        10878|     inherit (pkgs) cmake ninja qt5;

       … while calling 'callPackageWith'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/customisation.nix:139:35:

          138|   */
          139|   callPackageWith = autoArgs: fn: args:
             |                                   ^
          140|     let

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/trivial.nix:430:7:

          429|     { # TODO: Should we add call-time "type" checking like built in?
          430|       __functor = self: f;
             |       ^
          431|       __functionArgs = args;

       … while calling anonymous lambda

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/customisation.nix:80:6:

           79|   makeOverridable = f: lib.setFunctionArgs
           80|     (origArgs: let
             |      ^
           81|       result = f origArgs;

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/customisation.nix:81:16:

           80|     (origArgs: let
           81|       result = f origArgs;
             |                ^
           82|

       … while calling anonymous lambda

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/python-modules/pyside2/default.nix:1:1:

            1| { python
             | ^
            2| , pythonAtLeast

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/python-modules/pyside2/default.nix:17:1:

           16| # "There are no plans to support Python versions > 3.10 in the 5.15 branch."
           17| disabledIf (pythonAtLeast "3.11") (
             | ^
           18| stdenv.mkDerivation rec {

       … while calling 'disabledIf'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/interpreters/python/python-packages-base.nix:92:19:

           91|
           92|   disabledIf = x: drv: if x then disabled drv else drv;
             |                   ^
           93|

       … from call site

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/interpreters/python/python-packages-base.nix:92:34:

           91|
           92|   disabledIf = x: drv: if x then disabled drv else drv;
             |                                  ^
           93|

       … while calling 'disabled'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/pkgs/development/interpreters/python/python-packages-base.nix:90:14:

           89|
           90|   disabled = drv: throw "${removePythonPrefix (drv.pname or drv.name)} not supported for interpreter ${python.executable}";
             |              ^
           91|

       error: pyside2 not supported for interpreter python3.11
```

### Additional context

Problem stated pretty clearly in the build log, the recent change to using python 3.11 by default in nixos unstable means that the build input of freeCAD, pyside2, fails to build. 

### Notify maintainers

@viric 
@gebner 
@AndersonTorres 

### Metadata

Please run `nix shell nixpkgs#nix-info -c "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix shell nixpkgs#nix-info -c "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.7, NixOS, 23.11 (Tapir), 23.11.20231024.8efd5d1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.1`
 - channels(root): `"nixos-23.05, nixpkgs-unstable"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by W-Y-R** [comment]
I've recently updated my flake and now I'm having trouble rebuilding my system due to the same error. I'm not entirely sure where it is coming from.
```
error:
       … while calling the 'head' builtin

         at /nix/store/35dcag44a0ymww0vy0s4jjgxwpv9g62d-source/lib/attrsets.nix:850:11:

          849|         || pred here (elemAt values 1) (head values) then
          850|           head values
             |           ^
          851|         else

       … while evaluating the attribute 'value'

         at /nix/store/35dcag44a0ymww0vy0s4jjgxwpv9g62d-source/lib/modules.nix:807:9:

          806|     in warnDeprecation opt //
          807|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          808|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: pyside2 not supported for interpreter python3.11
```

**Comment by lilyinstarlight** [comment]
This should have been fixed by #263599, which is in the nixos-unstable and nixpkgs-unstable channels now

Feel free to reopen if you are still having an issue after updating nixpkgs!
